packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -505,6 +620,271 @@ class NavigatorObserver {
+      newPageRouteHistory: newPageRouteHistory,
+      locationToExitingPageRoute: locationToExitingPageRoute,
+      pageRouteToPagelessRoutes: pageRouteToPagelessRoutes,
+    );
+    // Verifies the integrity after the decisions has made.
+    //
+    // Here are the rules:
+    // - All the entering routes in newPageRouteHistory must either be pushed or added.
+    // - All the exiting routes in locationToExitingPageRoute must either be popped or removed.
+    // - All the pageless routes must either be popped or removed.
+    // - All the entering routes in the result must preserve the same order as
+    //   the entering routes in newPageRouteHistory and contain all exiting
+    //   routes.
+    //     ex:
+    //
+    //     newPageRouteHistory = [A, B, C]
+    //
+    //     locationToExitingPageRoute = {A -> D, C -> E}
+    //
+    //     results = [A, B ,C ,D ,E] is valid
+    //     results = [D, A, B ,C ,E] is also valid because exiting route can be
+    //     inserted in any place
+    //
+    //     results = [B, A ,C ,D ,E] is invalid because B must be after A.
+    //     results = [B, A ,C ,E] is invalid because results must include D and E.
+    assert(() {
+      final List<StageableRoute> resultsToVerify = results.toList(growable: false);
+      final Set<StageableRoute> exitingPageRoutes = locationToExitingPageRoute.values.toSet();
+      // Firstly, verify all exiting route has been marked.
+      for (final StageableRoute exitingPageRoute in exitingPageRoutes) {
+        assert(!exitingPageRoute._debugWaitingForExitDecision);
+        if (pageRouteToPagelessRoutes.containsKey(exitingPageRoute)) {
+          for (final StageableRoute pagelessRoute in pageRouteToPagelessRoutes[exitingPageRoute]) {
+            assert(!pagelessRoute._debugWaitingForExitDecision);
+          }
+        }
+      }
+      // Secondly, verifies the order of results matches the newPageRouteHistory
+      // and contains all the exiting routes.
+      int indexOfNextRouteInNewHistory = 0;
+
+      for (final _RouteEntry routeEntry in resultsToVerify.cast<_RouteEntry>()) {
+        assert(routeEntry != null);
+        assert(!routeEntry.isEntering && !routeEntry._debugWaitingForExitDecision);
+        if (
+          indexOfNextRouteInNewHistory >= newPageRouteHistory.length ||
+          routeEntry != newPageRouteHistory[indexOfNextRouteInNewHistory]
+        ) {
+          assert(exitingPageRoutes.contains(routeEntry));
+          exitingPageRoutes.remove(routeEntry);
+        } else {
+          indexOfNextRouteInNewHistory += 1;
+        }
+      }
+
+      assert(
+       indexOfNextRouteInNewHistory == newPageRouteHistory.length &&
+       exitingPageRoutes.isEmpty
+      );
+      return true;
+    }());
+
+    return results;
+  }
+
+  /// A method that will be called by the [Navigator] for deciding how route
+  /// transitions in or out of the screen.
+  ///
+  /// The [newPageRouteHistory] represents the new page-based route after the
```

Comment:
The `newPageRouteHistory` list contains all page-based routes in order that will be on the [Navigator]'s history stack after this update completes.